beer-garden/functionary
# #91 - Only consume messages if a worker is available, added listener unit tests, removed redis as a broker

This PR closes #91

## Introduced Changes
- Removed the `Filesystem` broker from the `runner` because it does not support monitoring, is poorly documented, and does not appear to have first-class support like Redis or RabbitMQ
- Listener will not consumer another message from the queue unless a worker is available to consume the message
  - A worker is available if one of its spawned sub-processes is not currently executing a task. This can be determined by comparing the tasks assigned to the worker with the worker's concurrency setting. If the number of assigned tasks is less than the worker's number of concurrent processes, then the worker has an available process to consume another message from the queue.
- Added unit tests to the listener to test the functionality of the busy/wait logic
- Removed Redis from Functionary in favor of just using RabbitMQ as the message broker
- Added `pytest` and `pytest-mock` to the `requirements-dev.txt` for the `runner`

## Testing
- To test the added changes to the `runner.listener`, run the unit tests found in the `runner/tests/`
  - Open to suggestions on how I should change the structure of the unit tests
- Start Functionary, and then inspect any of the logs for containers that have a Celery worker inside of them
  - You should see 3 total workers join a party, since they are all connecting to the same message broker
  - Publish a package to Functionary. You should still see the proper worker consume the task and build the package
  - Run a Functionary task. You should still see the proper worker consume the task and publish the results


## Celery Worker isolation decision that will need to be made
Since all of the Celery workers are connecting to the same messaging broker and `vhost`, they are all visible to each other, hence why the Celery logs show that another worker joined the party. If we want to separate out the visibility of the Celery workers from each other, I believe we could use [RabbitMQ's `vhost` setting](https://www.rabbitmq.com/vhosts.html), such that the `runner` can connect to one or many separate `vhost`s, and let the functionary and functionary builder workers live on a separate `vhost`. I believe this would prevent the Celery workers from being visible to one another, unless they are connected to the same `vhost`. I think this could also help us when we want to implement private `runner`s, since we could just spawn another `vhost` for that specific runner via RabbitMQ's API.
